### ISSUE TITLE
Adding support for roles

### ISSUE THREAD

**Issue by ocram** [delight-im/PHP-Auth#13]
We should add support for user roles to this library soon. A list of available roles can be hard-coded and could look like this:

```php
final class Role {

	const ADMIN = 1;
	const AUTHOR = 2;
	const COLLABORATOR = 4;
	const CONSULTANT = 8;
	const CONSUMER = 16;
	const CONTRIBUTOR = 32;
	const COORDINATOR = 64;
	const CREATOR = 128;
	const DEVELOPER = 256;
	const DIRECTOR = 512;
	const EDITOR = 1024;
	const EMPLOYEE = 2048;
	const MAINTAINER = 4096;
	const MANAGER = 8192;
	const MODERATOR = 16384;
	const PUBLISHER = 32768;
	const REVIEWER = 65536;
	const SUBSCRIBER = 131072;
	const SUPER_ADMIN = 262144;
	const SUPER_EDITOR = 524288;
	const SUPER_MODERATOR = 1048576;
	const TRANSLATOR = 2097152;

	private function __construct() {}

}
```

You can then use any of these roles and ignore those that you don't need, and you can even combine these roles as you like. Thus a user may have none of these rules, one role or any other arbitrary subset.

Although a fixed set of roles is hard-coded in this implementation, the set of roles should include identifiers for most use cases. And if you really can't work with these roles, you can alias them using your own role identifiers:

```php
final class MyRole {

	const CUSTOMER_SERVICE_AGENT = Role::REVIEWER;
	const FINANCIAL_DIRECTOR = Role::COORDINATOR;

	private function __construct() {}

}
```

That, for example, would allow you to use `MyRole::FINANCIAL_DIRECTOR` instead of `Role::COORDINATOR`.

Checking for roles that the current user has might then be done like this:

```php
/** @var Auth $auth */

if ($auth->hasRole(Role::ADMIN)) {
	// ...
}

// or

if ($auth->hasAnyRole(Role::MODERATOR, Role::SUPER_MODERATOR, Role::ADMIN)) {
	// ...
}

// or

if ($auth->hasAllRoles(Role::TRANSLATOR, Role::REVIEWER)) {
	// ...
}
```

This feature would not include support for permissions, capabilities, access rights or privileges. These could be added with custom solutions in user code or one could go entirely without them. As can be seen above, it's possible to implement access restrictions using roles only. The privileges are then encoded in the conditions that specify the required roles.

Since the roles require database changes (at least one new column in the `users` table) that must be applied manually, this is a breaking change that can only be in a new major release. The next one would currently be `v5.0.0`.

Feedback, requests for additions/changes and criticism is welcome!

**Comment by Furentes** [comment]
May I ask when this feature will be available?

**Comment by ocram** [comment]
Well, the idea was that we would gather some more feedback on this.

We better implement something that everyone finds helpful *late* than implement something that turns out to be the wrong feature *early*. The worst case would be if it later turned out that we did the feature the wrong way.

As you seem to be interested in this feature, why not discuss the implementation suggested above. Specifically, the proposal does include groups or roles only, but no individual rights or permissions. Are you fine with that? Additionally, it does not allow for custom roles, but only provides a hard-coded set of pre-defined roles. Is that okay as well? (As shown above, it would be possible to alias the roles with custom names, though.)

**Comment by Furentes** [comment]
Sorry for the late response.

Rather than hard coding roles, how about a simple permission system like the one from TeamSpeak 3?
Because I needed something like that, I created a **users_permissions** table with the columns **user** and **value**. **User** is the ID from the user and **value** is the permission value.

Let's say i have a page for administrators to manage accounts, i run a function called **hasPermission(75)**, which returns true if the user has a value the same or greater than the given one. Now if the user has a permission value over 75, he is able to access the site. Otherwise, it returns false and i can decide what happens (e.g. redirect). 

Like that, everyone is able to "set his own roles", without having hard coded roles which can not really be changed. Roles could be defined like that:
```
define("OPERATOR", 60);
define("ADMINISTRATOR", 70);
```
then the function looks like:
`hasPermission(OPERATOR);`

(Sorry for any spelling mistakes, english isn't my native language.)

**Comment by ocram** [comment]
Thanks for explaining how you have implemented this!

Since your solution does not differentiate between "permissions", "groups" and "roles" – they're all the same, just as in the idea outlined above – our two ideas seem to be more similar than one might think at first.

In order to represent rights that multiple roles have, your solution lets roles with higher numbers do everything that roles with lower numbers can do. That's fine, but it doesn't allow a "manager" to do *some* things the "editor" can do and *some* things the "moderator" can do.

In the solution above, this is implemented in a way that is a little more flexible: First, some right can be granted to multiple roles, as in your case. But these roles can be chosen accurately – "higher" roles don't inherit *everything* that "lower" roles can do. Second, a user may have multiple roles instead of only one.

That being said, your solution definitely isn't a bad one and I'm sure it works fine for your use case. But for this library, I haven't really decided yet how to implement this so that it works for as many people as possible.

**Comment by fdiengdoh** [comment]
Just a thought.
I've been trying to use your script in one of my project. I've currently used my own authentication system, which I find its not secure. But, in my system, I've used three roles, `admin`, `teacher` and `part_time`. Which I would love to see implemented in this script of yours. It seems @Furentes solution is a more flexible one that we can tweak to our own needs rather than hard code it.

**Comment by ocram** [comment]
Thanks for your feedback, @fdiengdoh!

I see that @Furentes' solution appears to be more flexible at first. But that seemingly higher flexibility is just the freedom to use custom labels for the roles, isn't it? That advantage is made up for in the first solution proposed above by the possibility to alias any roles with custom names (see above).

If it were only that small difference, I'd say it doesn't matter. But in @Furentes' solution, there are two things that aren't really perfect for a general-purpose solution in my opinion:

 * Users can't have more than a single role.
 * Roles with higher numbers always inherit *all* rights that *every* role with a lower number has.

**Comment by fdiengdoh** [comment]
Thanks for the clarification, I'm not an expert in these stuff though. Looking forward for a proper implementation in up coming updates. 

Just a thought, how long will it take to add this enhancement? If its not going to be long then, I can wait for it, or may have to do something in my project till the enhancement is available.

**Comment by Furentes** [comment]
Hey, @fdiengdoh! As @ocram already stated:

>  Well, the idea was that we would gather some more feedback on this.

So this **could** take a while.

Now I'm getting the exact problem @ocram stated, with my own implementation. 

> Users can't have more than a single role

So I got an idea.
Back in the "old days", where i played Minecraft, I used a plugin called [PermissionEX](https://github.com/PEXPlugins/PermissionsEx/wiki/Pex-basics) to handle groups and permissions on my server. Now that I think of it, we could use a similar approach. 

Let's say we can create on our own groups in in an table manually or with an function to insert them automatically. Also there is an permission table, where name and a value (if needed) is stored. With an function, we can check, if someone has specific groups or specific permissions. (e.g. `hasGroup($array)` or `hasPermission($array)`)

The §array could look like this:
(I created a JSON and used json_decode($json, true) to get this array)
```
Array
(
    [permissions] => Array
        (
            [0] => Array
                (
                    [name] => edit.blog.entry
                    [value] => 10
                )

            [1] => Array
                (
                    [name] => edit.news.entry
                    [value] => 25
                )

            [2] => Array
                (
                    [name] => edit.users
                    [value] => 12
                )
        )
)
```

Like that everyone is able to create his own groups and permission, and also give them a custom name.

**Comment by ocram** [comment]
Thanks for your additional feedback, @fdiengdoh and @Furentes!

I have evaluated the possible solutions again, checked them against our requirements and examined how easy to use and flexible they are. Having done that, I decided I will implement the first solution that I proposed here. This will be done shortly.

Some people put a layer of permissions, activities, capabilities, abilities, access rights or privileges on top of the roles. You then don't check for role or group membership in your business logic, but you check for individual permissions. This has some advantages: If you want to know who can edit articles, for example, you don't have to check your business logic in various places, but you only have to look where the permission is defined. If you want to *change* who can edit an article, you only have to do this in one single place, not throughout your whole code base. But this comes with slightly more overhead when implementing the access restrictions the first time. For lots of use cases, and especially for smaller projects, the advantages are not really significant, however.

So that's why we go without an additional system of permissions first. If you need that additional abstraction, you can implement it in your own code, e.g. like this:

```php
function canEditArticle(\Delight\Auth\Auth $auth) {
	return $auth->hasAnyRole(
		\Delight\Auth\Role::MODERATOR,
		\Delight\Auth\Role::SUPER_MODERATOR,
		\Delight\Auth\Role::ADMIN,
		\Delight\Auth\Role::SUPER_ADMIN
	);
}

// ...

if (canEditArticle($app->auth())) {
	// ...
}

// ...

if (canEditArticle($app->auth())) {
	// ...
}

// ...

if (canEditArticle($app->auth())) {
	// ...
}
```

Development has started here: https://github.com/delight-im/PHP-Auth/tree/roles

**Comment by ocram** [comment]
Implemented in:

 * https://github.com/delight-im/PHP-Auth/commit/63734fc5ee90cca27fc95a2b233c02822beb6793
 * https://github.com/delight-im/PHP-Auth/commit/8c2c32f9dc362155d6a21fa959a9806a1f7f78e8
 * https://github.com/delight-im/PHP-Auth/commit/6cf955ed52b8420b83e2021187a3a500be1e0eaf
 * https://github.com/delight-im/PHP-Auth/commit/b577322939be4cffd58a5eea87114cb4d4fc82b7
 * https://github.com/delight-im/PHP-Auth/commit/d6bc8c64921fb5a3c49012cfd980c3c803b155e0
 * https://github.com/delight-im/PHP-Auth/commit/2278b86fbaafd6555cdfb53fdb3a47140d1a8b28
 * https://github.com/delight-im/PHP-Auth/commit/c25b74d405bb9d2b0b296e6e2ef8a59a9bd6039d
 * https://github.com/delight-im/PHP-Auth/commit/23b172055bc8e2c95e71d77d3542185d30094eea
 * https://github.com/delight-im/PHP-Auth/commit/8a1140a485b0ac6dfb4718f09a13ba929081c738
 * https://github.com/delight-im/PHP-Auth/commit/9c63c30cd9bc928632092ddfeb25af298f6a85a1
 * https://github.com/delight-im/PHP-Auth/commit/f2074e153796cb2bac9bd246c91063161bcd4701
 * https://github.com/delight-im/PHP-Auth/commit/b7e6ca6dee04a0fd3d58e5946b80327601becc23
 * https://github.com/delight-im/PHP-Auth/commit/1a4041ea6028cbf9a25dad9d5b3bce90a4a530f3
 * https://github.com/delight-im/PHP-Auth/commit/8ff4242f8ffc8f3ecbdbdacd4f687d3c4071272f
 * https://github.com/delight-im/PHP-Auth/commit/bc44a08b1b3a71ca5fb356d0036fbcb68660bd93
 * https://github.com/delight-im/PHP-Auth/commit/cd3469c137ff1938b7d7e70293e9db5112323ba5
 * https://github.com/delight-im/PHP-Auth/commit/7b6287a7dc47e7f2d6da9016053536be8c6e9016
 * https://github.com/delight-im/PHP-Auth/commit/0af55ad77c2a8ff3b2ffaec423091f60d71b4574
 * https://github.com/delight-im/PHP-Auth/commit/b93d9616d00914745acb225dd2ef0872e6c51001

This is available in a new major version, `v6.0.0`, which comes with a few breaking changes. For a guide on how to upgrade, please see the migration notes in `Migration.md`.
